### DISCUSSION TITLE
How to set Customer ID user field of a contact item?

### DISCUSSION THREAD

**Discussion by Tre-Seibert** [ecederstrand/exchangelib#1240] [Q&A] [answered]
I am trying to create new contacts and add a value for the 'Customer ID' field. Can someone please give me some insight as to how to accomplish this.

```
`contact = Contact(
    account=account,
    folder=fTBClients,
    given_name="Test",
    surname="Test Contact",
    display_name="Test Contact",
    company_name= "sample Corporation",
    customer_id = "SAMP"
)

contact.save()
print("New contact made")`
```

`    contact = Contact(
              ^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/exchangelib/items/item.py", line 123, in __init__
    super().__init__(**kwargs)
  File "/usr/local/lib/python3.11/site-packages/exchangelib/items/base.py", line 166, in __init__
    super().__init__(**kwargs)
  File "/usr/local/lib/python3.11/site-packages/exchangelib/properties.py", line 1475, in __init__
    super().__init__(**kwargs)
  File "/usr/local/lib/python3.11/site-packages/exchangelib/properties.py", line 270, in __init__
    raise AttributeError(f"{sorted(kwargs.keys())!r} are invalid kwargs for this class")
AttributeError: ['customer_id'] are invalid kwargs for this class`

<img width="1066" alt="Screenshot 2023-10-19 at 11 55 04 PM" src="https://github.com/ecederstrand/exchangelib/assets/63426280/50c129b3-d8b7-4997-a929-dc08a121e12c">

**Comment by ecederstrand** [comment]
I don't see "Customer ID" mentioned anywhere in MSDN documentation: https://learn.microsoft.com/en-us/exchange/client-developer/web-service-reference/contact

Do you have any resources indicating that this is a standard field on Contact in Exchange?

**Reply by Tre-Seibert** [reply]
Under the 'created' field there's a field called "Customer ID", that's the field I want to manipulate. It's a contact item field in Outlook 16

**Reply by Tre-Seibert** [reply]
https://learn.microsoft.com/en-us/office/vba/api/outlook.contactitem.customerid

**Reply by ecederstrand** [reply]
Ok, I just don't see any indication that it's supported in via the EWS API. But you can always try to add a field definition to this field to the `Contact` class (see https://github.com/ecederstrand/exchangelib/blob/6a8289a411b771a34f6e46642fe9b486367c3860/exchangelib/items/contact.py#L40) and run your initial code example again.

**Reply by Tre-Seibert** [reply]
Tried inserting this line into the Contact class: `customer_id = TextField(field_uri="contacts:CustomerID")`. The code runs without error and if I print(contact.customer_id), the value is printed. However, it does not show up in outlook. Any other ideas? Thank you for your response.

**Reply by ecederstrand** [reply]
There's an answer on StackOverflow hinting that this field is in fact an extended property: https://stackoverflow.com/a/16726773/219640 You could try creating an extended property in exchangelib with those property definitions and see if that does the trick.

**Reply by Tre-Seibert** [reply]
Worked thank you!

**Comment by Tre-Seibert** [comment, marked as answer]
Working code in this test env, see property tag id found at https://msopenspecs.azureedge.net/files/MS-OXPROPS/%5bMS-OXPROPS%5d.pdf:

```
# Define the 'CustomerID' extended property
class customerID(ExtendedProperty):
    property_type = "String"
    property_tag= 0x3A4A

# Register the 'CustomerID' property on the Contact item type
Contact.register('customerIDProperty', customerID)
```
